### ISSUE TITLE
Introduce an error-code based control mechanism for preventing access to the checkout block

### ISSUE THREAD

**Issue by manospsyx** [woothemes/woocommerce#42648]
## Is your feature request related to a problem? Please describe.

Currently, the Checkout Block:

1. Does not convert the `errors` in cart responses to notices.
2. Does not provide a formal way to developers who need to control access to the Checkout Block.

This issue is about (2). A separate issue has been opened to discuss (1).

An erroneous cart state might be a result of:

- <strong>Cart item errors</strong> -- these might be triggered by the presence of out of stock items, violations of quantity constraints, or invalid product configurations. Add-to-cart validation is necessary to prevent these errors, but it is not sufficient as stock, quantity constraints, or product configuration constraints may change while a product is in the cart.
- <strong>Coupon errors</strong> -- these might be triggered by changes in cart state that may render a coupon invalid.
- <strong>Errors triggered by third-party functionality</strong> -- there's a myriad of plugins that implement different types of business logic, such as our own <a href="https://somewherewarmattic.wordpress.com/2022/02/01/conditional-shipping-and-payments-cart-and-checkout-blocks-integration-phase-2/">[Conditional Shipping and Payments](https://somewherewarmattic.wordpress.com/2022/02/01/conditional-shipping-and-payments-cart-and-checkout-blocks-integration-phase-2/)</a>. Other examples include gift cards, bundles, bookings, etc.

Certain types of errors create the need to [prevent users from accessing the checkout block](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5889).

However, other types of errors might be **designed to be rendered inside the Checkout Block**. For example, if the erroneous state is triggered by a shipping address change, or a shipping method change (e.g. "allow Local Pickup only if the shipping postcode is 17235"), then blocking access to the Checkout Block upon entering the error state could end up creating a user experience issue.

Given that the <code>errors</code> field is the single source of truth for <strong>all erroneous cart states</strong>, developers must be able to define precisely which error codes should prevent access to the Checkout Block.

## Describe the solution you'd like

Currently, this can be forced by throwing a <code>RouteException</code> during the draft order creation/update process. A possibly better solution would be to <strong>let developers define specific error codes that control this behavior</strong>. 

One way to do this would be to include a filterable `wcSettings` [field for this purpose](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/44b633957a5356414b93903b568aca81d83e3d6b/src/Assets/AssetDataRegistry.php#L356). 

Depending on the `errors` returned by the API:

- The Cart Block will prevent access to the Checkout Block when clicking Proceed to Checkout.
- The Checkout Block will render in a "warning" state, and prompt users to resolve their cart errors to proceed.

For example, when the `errors` field in cart route responses contain a `woocommerce-blocks-product-out-of-stock`, `woocommerce-blocks-product-not-purchasable`, `woocommerce-blocks-product-partially-out-of-stock` or `woocommerce-blocks-too-many-of-product-in-cart` error code, the Cart Block could trigger this behavior and disable the **Proceed to Checkout** button. A tooltip or alert (worse) could be used to direct the user's attention to the rendered notices.

## Additional context

Refer to [this post](https://somewherewarmattic.wordpress.com/2022/02/18/extensibility-review-error-handling-and-validation-in-the-block-based-checkout/).

**Comment by ralucaStan** [comment]
After https://github.com/woocommerce/woocommerce-blocks/pull/8162 the Checkout block converts the errors into notices. 

> Depending on the errors returned by the API:
> The Cart Block will prevent access to the Checkout Block when clicking Proceed to Checkout.
> The Checkout Block will render in a "warning" state, and prompt users to resolve their cart errors to proceed.


If there is an error where the checkout can't proceed this is the current behavior:

In the Cart block, there is an error notice and the button is not disabled.
![Screenshot 2023-02-02 at 15 06 32](https://user-images.githubusercontent.com/1628454/217258681-c9dd614c-250a-4aef-bf6c-0a3cbcdbb5d0.png)

In the Checkout block, the user can see this state:
![Screenshot 2023-02-02 at 14 59 55](https://user-images.githubusercontent.com/1628454/217258667-dcd4c869-a118-405b-84b2-6532803a0570.png)

It is possible to disable the Proceed to checkout button in the Cart block when we have an error in the cart store. But the checkout page can also be accessed by direct link, so we can't guarantee that the page is not accessed.

@elizaan36 do you think we should disable the Proceed to checkout block in the Cart block in this case?

**Comment by mikejolley** [comment]
Moving this back to _next_ until it has feedback.

**Comment by elizaan36** [comment]
Yes, I think it makes sense to disable the Proceed to checkout button in this case. It's quite confusing for the shopper to see a Cart error on a page where they can't take action on it.

AFAIU when the user tries to visit the Checkout from the navigation, they would just be redirected to the cart with the notice "Checkout is not available while your cart is empty." In this case, could they be redirected to the cart with this notice?

**Comment by opr** [comment]
Hi @elizaan36 and @ralucaStan - I don't think we should actually disable the button. We should instead allow the shopper to click it, but if there are errors when they do, it should just highlight those errors by scrolling.

By preventing them from clicking, they could be frustrated by not understanding what's wrong with it, and why the button doesn't work. Allowing the click, then highlighting the error will make it easier for them to know why they can't proceed, and give them the opportunity to fix it, without needing to figure it out themselves.

I appreciate it's only a single thread, but a lot of folks agree that disabled buttons are a poor UX for which better alternatives exist. https://twitter.com/adambsilver/status/1663588645020762127

**Comment by elizaan36** [comment]
hey @opr I think this was an old one before we decided against disabling the button in later threads.

> I don't think we should actually disable the button

+1

**Comment by pierorocca** [comment]
I've taken a look and marked this ready for the engineering team to pick up.
